Incompatibility between ff4j-store-mongodb-v3 and Spring boot 2.3.x

Hello everyone!

I am working on Spring Boot version migration to 2.3.x but I realized that I am not being able to use ff4j-store-mongodb-v3 and ff4j 1.8.9 because the class FeatureStoreMongo imports MongoClient from package com.mongodb instead com.mongodb.client used by Spring Data new version.

My pom is:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>com.progtiago</groupId>
  <artifactId>mp-image</artifactId>
  <version>1.0-SNAPSHOT</version>

  <parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>
    <version>2.3.3.RELEASE</version>
    <relativePath/>
  </parent>

  <properties>
    <java.version>1.8</java.version>
    <spring.cloud.kubernetes.version>1.0.0.RC2</spring.cloud.kubernetes.version>
    <spring-cloud-starter-openfeign.version>2.1.5.RELEASE</spring-cloud-starter-openfeign.version>
    <springfox.swagger.version>2.9.2</springfox.swagger.version>
    <ff4j.version>1.8.9</ff4j.version>
    <commons.io.version>2.6</commons.io.version>
    <commons.net.version>3.6</commons.net.version>
    <commons-lang.version>3.9</commons-lang.version>
    <aws-java-sdk>1.11.517</aws-java-sdk>
    <fixture-factory.version>3.1.0</fixture-factory.version>
    <nal-logging.version>1.7</nal-logging.version>
    <newrelic-api.version>4.11.0</newrelic-api.version>
    <commons-collections4.version>4.2</commons-collections4.version>
    <twelvemonkeys-imageio.version>3.5</twelvemonkeys-imageio.version>
    <imgscalr-lib.version>4.2</imgscalr-lib.version>
    <pitest-plugin.version>1.4.6</pitest-plugin.version>
    <powermock.version>2.0.0</powermock.version>
    <mockito-all.version>1.10.19</mockito-all.version>
    <!-- Container Tests -->
    <testcontainers.version>1.12.4</testcontainers.version>
    <mockserver.version>1.12.0</mockserver.version>
    <mockserver-client-java.version>5.6.1</mockserver-client-java.version>

    <!-- Sonar -->
    <sonar.sources>src/main/java</sonar.sources>
    <sonar.tests>src/test/</sonar.tests>
    <sonar.sourceEncoding>UTF-8</sonar.sourceEncoding>
    <sonar.jacoco.reportPath>target/jacoco.exec</sonar.jacoco.reportPath>
    <sonar.java.binaries>target/classes</sonar.java.binaries>

  </properties>

  <dependencies>
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-starter-kubernetes</artifactId>
      <version>${spring.cloud.kubernetes.version}</version>
    </dependency>
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-starter-kubernetes-config</artifactId>
      <version>${spring.cloud.kubernetes.version}</version>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-web</artifactId>
      <exclusions>
        <exclusion>
          <groupId>org.springframework.boot</groupId>
          <artifactId>spring-boot-starter-tomcat</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-undertow</artifactId>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-actuator</artifactId>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-data-mongodb</artifactId>
    </dependency>
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-amqp</artifactId>
    </dependency>
    <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-starter-openfeign</artifactId>
      <version>${spring-cloud-starter-openfeign.version}</version>
    </dependency>

    <!-- Image Scaling -->
    <dependency>
      <groupId>org.imgscalr</groupId>
      <artifactId>imgscalr-lib</artifactId>
      <version>${imgscalr-lib.version}</version>
    </dependency>

    <dependency>
      <groupId>com.twelvemonkeys.imageio</groupId>
      <artifactId>imageio-jpeg</artifactId>
      <version>${twelvemonkeys-imageio.version}</version>
    </dependency>

    <!-- Lombok -->
    <dependency>
      <groupId>org.projectlombok</groupId>
      <artifactId>lombok</artifactId>
      <version>${lombok.version}</version>
      <scope>provided</scope>
    </dependency>

    <!-- Springfox Swagger -->
    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-swagger2</artifactId>
      <version>${springfox.swagger.version}</version>
    </dependency>
    <dependency>
      <groupId>io.springfox</groupId>
      <artifactId>springfox-swagger-ui</artifactId>
      <version>${springfox.swagger.version}</version>
    </dependency>

    <!-- Ff4j -->
    <dependency>
      <groupId>org.ff4j</groupId>
      <artifactId>ff4j-core</artifactId>
      <version>${ff4j.version}</version>
    </dependency>
    <dependency>
      <groupId>org.ff4j</groupId>
      <artifactId>ff4j-web</artifactId>
      <version>${ff4j.version}</version>
      <exclusions>
        <exclusion>
          <groupId>org.thymeleaf</groupId>
          <artifactId>thymeleaf</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.ff4j</groupId>
      <artifactId>ff4j-store-mongodb-v3</artifactId>
      <version>${ff4j.version}</version>
    </dependency>
    <dependency>
      <groupId>org.ff4j</groupId>
      <artifactId>ff4j-aop</artifactId>
      <version>${ff4j.version}</version>
    </dependency>

    <!-- NewRelic -->
    <dependency>
      <groupId>com.newrelic.agent.java</groupId>
      <artifactId>newrelic-api</artifactId>
      <version>${newrelic-api.version}</version>
    </dependency>

    <dependency>
      <groupId>com.fasterxml.jackson.datatype</groupId>
      <artifactId>jackson-datatype-jsr310</artifactId>
    </dependency>

    <!-- Aws Sdk -->
    <dependency>
      <groupId>com.amazonaws</groupId>
      <artifactId>aws-java-sdk</artifactId>
      <version>${aws-java-sdk}</version>
    </dependency>

    <!-- Commons -->
    <dependency>
      <groupId>commons-io</groupId>
      <artifactId>commons-io</artifactId>
      <version>${commons.io.version}</version>
    </dependency>

    <dependency>
      <groupId>commons-net</groupId>
      <artifactId>commons-net</artifactId>
      <version>${commons.net.version}</version>
    </dependency>

    <dependency>
      <groupId>org.apache.commons</groupId>
      <artifactId>commons-collections4</artifactId>
      <version>${commons-collections4.version}</version>
    </dependency>

    <dependency>
      <groupId>org.apache.commons</groupId>
      <artifactId>commons-lang3</artifactId>
      <version>${commons-lang.version}</version>
    </dependency>

    <!-- Jolokia -->
    <dependency>
      <groupId>org.jolokia</groupId>
      <artifactId>jolokia-spring</artifactId>
      <classifier>plugin</classifier>
      <version>${jolokia.version}</version>
    </dependency>

    <!-- Test -->
    <dependency>
      <groupId>br.com.six2six</groupId>
      <artifactId>fixture-factory</artifactId>
      <version>${fixture-factory.version}</version>
      <scope>test</scope>
    </dependency>

    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-test</artifactId>
      <scope>test</scope>
    </dependency>

    <dependency>
      <groupId>org.powermock</groupId>
      <artifactId>powermock-reflect</artifactId>
      <version>${powermock.version}</version>
      <scope>test</scope>
    </dependency>

    <!--  Test container  -->
    <dependency>
      <groupId>org.testcontainers</groupId>
      <artifactId>testcontainers</artifactId>
      <version>${testcontainers.version}</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.mockito</groupId>
      <artifactId>mockito-all</artifactId>
      <version>${mockito-all.version}</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.testcontainers</groupId>
      <artifactId>mockserver</artifactId>
      <version>${mockserver.version}</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.mock-server</groupId>
      <artifactId>mockserver-client-java</artifactId>
      <version>${mockserver-client-java.version}</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
  
  <repositories>

  </repositories>
</project>
```
My FF4J Config class is:

```
package com.progtiago.image.configurations.ff4j;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoDatabase;
import java.util.Arrays;
import org.ff4j.FF4j;
import org.ff4j.core.Feature;
import org.ff4j.core.FeatureStore;
import org.ff4j.mongo.store.FeatureStoreMongo;
import org.ff4j.web.ApiConfig;
import org.ff4j.web.embedded.ConsoleServlet;
import org.springframework.boot.web.servlet.ServletRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.mongodb.core.MongoTemplate;

@Configuration
@ComponentScan(basePackages = "org.ff4j.aop")
public class FF4JConfiguration {
  private static final String DEFAULT_CONSOLE = "/ff4j-console/*";
  private static final String COLLECTION_NAME = "ff4j-features";

  @Bean
  public FF4j getFF4j(final FeatureStore featureStore) {
    final FF4j ff4j = new FF4j();
    ff4j.disableAlterBeanThrowInvocationTargetException();
    ff4j.setFeatureStore(featureStore);
    Arrays.stream(Features.values())
        .filter(feature -> !ff4j.getFeatureStore().exist(feature.getKey()))
        .forEach(feature -> createFeature(ff4j, feature));
    return ff4j;
  }

  @Bean
  public ConsoleServlet getFF4jServlet(final FF4j ff4j) {
    final ConsoleServlet ff4jConsoleServlet = new ConsoleServlet();
    ff4jConsoleServlet.setFf4j(ff4j);
    return ff4jConsoleServlet;
  }

  @Bean
  public ApiConfig getApiConfig(final FF4j ff4j) {
    final ApiConfig apiConfig = new ApiConfig();
    apiConfig.setAuthenticate(false);
    apiConfig.setAutorize(false);
    apiConfig.setFF4j(ff4j);
    return apiConfig;
  }

  @Bean
  public ServletRegistrationBean servletRegistrationBeanToFF4j(final FF4j ff4j) {
    return new ServletRegistrationBean(getFF4jServlet(ff4j), DEFAULT_CONSOLE);
  }

  @Bean
  public FeatureStore featureStore(
      final MongoClient mongoClient, final MongoTemplate mongoTemplate) {
    final MongoDatabase mongoDatabase = mongoClient.getDatabase(mongoTemplate.getDb().getName());
    return new FeatureStoreMongo(mongoDatabase, COLLECTION_NAME);
  }

  private void createFeature(final FF4j ff4j, final Features feature) {
    final Feature fp = new Feature(feature.getKey(), feature.isDefaultValue());
    fp.setDescription(feature.getDescription());
    fp.setGroup(feature.getGroup());
    ff4j.createFeature(fp);
  }
}
```

When I try compile, the error is this:

ff4j/FF4JConfiguration.java:[60,12] cannot access com.mongodb.MongoClient
[ERROR]   class file for com.mongodb.MongoClient not found

It is because, as I have said, the class FeatureStoreMongo imports MongoClient from com.mongodb instead of com.mongodb.client.

My conclusion is that the ff4j-store-mongodb-v3 is incompatible with Spring Data 2.3.x. I am right or somebody has another opinion?

Thanks so much!

Not sure why but keep having issue on my machine with test container now.

```
22:20:58.989 ERROR ResourceReaper - Timeout out waiting for Ryuk. Ryuk's log:
2020/10/04 20:20:28 Pinging Docker...
panic: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

goroutine 1 [running]:
main.main()
	/go/src/github.com/testcontainers/moby-ryuk/main.go:35 +0xcf5
```

- Reset docker (settings,images,containers..)
- Reinstalling docker
- Testing multiple version of test containers.

Will ignore the tests for now to be able to deliver. I know this is OK as travis is Happy.

--- a/ff4j-store-mongodb-v3/src/main/java/org/ff4j/mongo/store/FeatureStoreMongo.java
+++ b/ff4j-store-mongodb-v3/src/main/java/org/ff4j/mongo/store/FeatureStoreMongo.java
@@ -199,7 +199,7 @@ public class FeatureStoreMongo extends AbstractFeatureStore {
     @Override
     public boolean exist(String featId) {
         Util.assertHasLength(featId);
-        return 1 == getFeaturesCollection().count(BUILDER.getFeatUid(featId));
+        return 1 == getFeaturesCollection().countDocuments(BUILDER.getFeatUid(featId));
     }
 
     /** {@inheritDoc} */
@@ -296,7 +296,7 @@ public class FeatureStoreMongo extends AbstractFeatureStore {
         if (groupName == null || groupName.isEmpty()) {
             throw new IllegalArgumentException(GROUPNAME_CANNOT_BE_NULL_NOR_EMPTY);
         }
-        return getFeaturesCollection().count(BUILDER.getGroupName(groupName)) > 0;
+        return getFeaturesCollection().countDocuments(BUILDER.getGroupName(groupName)) > 0;
     }
 
     /** {@inheritDoc} */

--- a/ff4j-store-mongodb-v3/src/main/java/org/ff4j/mongo/store/PropertyStoreMongo.java
+++ b/ff4j-store-mongodb-v3/src/main/java/org/ff4j/mongo/store/PropertyStoreMongo.java
@@ -160,7 +160,7 @@ public class PropertyStoreMongo extends AbstractPropertyStore {
     /** {@inheritDoc} */
     public boolean existProperty(String name) {
         Util.assertHasLength(name);
-        return 1 == getPropertiesCollection().count(BUILDER.getName(name));
+        return 1 == getPropertiesCollection().countDocuments(BUILDER.getName(name));
     }
 
     /** {@inheritDoc} */

